iModelJsNodeAddon/JsInteropDgnDb.cpp

```diff
@@ -509,11 +509,20 @@ DgnDbStatus JsInterop::SimplifyElementGeometry(DgnDbR db, Napi::Object simplifyA
 /*---------------------------------------------------------------------------------**//**
 * @bsimethod
 +---------------+---------------+---------------+---------------+---------------+------*/
-void JsInterop::UpdateProjectExtents(DgnDbR dgndb, BeJsConst newExtents) {
+void JsInterop::UpdateProjectExtents(DgnDbR dgndb, BeJsConst newExtents, bool fromChangesetAppliedEvent) {
     auto& geolocation = dgndb.GeoLocation();
     AxisAlignedBox3d extents;
     extents.FromJson(newExtents);
-    geolocation.SetProjectExtents(extents);
+    if (fromChangesetAppliedEvent) {
+        // Check if our extents have actually changed. If not, return early.
+        if (geolocation.GetProjectExtents().IsEqual(extents))
+            return;
+        // set isValid to false so that we recalculate the ecefLocation in SetProjectExtents with these new extents.
+        auto ecefLocation = geolocation.GetEcefLocation();
+        ecefLocation.m_isValid = false;

```

Comment:
This seems wrong unless there is a GCS?

Reply:
Does ecefLocation's `m_isValid` get marked as true anywhere except when the GCS is valid? Im guessing if there is no GCS, this code will just mark m_isValid as false when it likely already was false and have no effect?

Now based off of your other comment, I am also wondering if I should not be checking the fromChangesetAPpliedEvent in thsi function at all (i.e. also mark eceflocation as invalid when updateprojectExtents is called with any new extents).

Reply:
You can have a valid ecef location without a gcs. The reason you can mark the ecef location invalid when there is a gcs is because the ecef location can be computed from the gcs (it's effectively redundant). No gcs though and the ecef location is all you have to geo-locate the model.

Reply:
Okay so that means the ecefLocation can be independently and explicitly set? I switched it to check that the DgnGCS is not nullptr before invalidating the eceflocation

Reply:
Yes, it can be independently set and like I said is redundant when there is a GCS, Why do you think you need to invalidate it at all?

Reply:
I assumed that the ecefLocation was dependent on the project extents because it uses the center of the projectExtents

```        
        auto projectCenter = extents.GetCenter();
        if (SUCCESS == LatLongFromXyz(cartographicOrigin, projectCenter))
            {
            m_ecefLocation = EcefLocation(cartographicOrigin, projectCenter, dgnGCS);
            }
```

My test in core also asserts that the ecefLocation changed after pulling a changeset which passes. Of course that is after I  invalidated the ecefLocation but it did change. The iModel i chose "mirukuru.ibim" also appears to have a GCS as there is a test which asserts that `ecefLocation for iModels` in core.

Reply:
> I assumed that the ecefLocation was dependent on the project extents
It's not, no reason to invalidate it...

That code snippet is what sets up the ecef transform from the GCS when it's not defined yet. It just wants a "meaningful" world location point to use to establish the ecef origin. In the absence of any other information, the project center is a decent choice.

Reply:
> Am I understanding that code snippet correctly that it will set a new ECEFLocation ?

Yes, it derives the ecef transform from the GCS, since it's redundant when a GCS is present as I keep saying. The GCS is what you always want to use to convert between spatial and cartographic when available anyway. The ecef location becomes increasingly inaccurate as you get farther from the ecef origin due to a design model being drawn on a plane and the earth being curved.

Reply:
Is there any documentation you could point me to so that I can understand this better?

Is setting the ecef transform just a formality when there is a GCS then? Why do we do it? I suppose so that any code which converts between spatial and cartographic has the option of using the ECEF transform? even though its not recommended to do so?

 What code is converting between spatial and cartographic? I had assumed maybe the GetGeoCoordsFromIModelCoords and GetIModelCoordsFromGeoCoords function but Im not sure thats right.

Reply:
The measure tools use IModelConnection.spatialToCartographic...

Reply:
Thanks, also clarified some things on call. Much appreciated. Feel free to resolve if comfortable